### PR TITLE
feat: physics character controller

### PR DESCRIPTION
**Pull request by yangfengzzz**
### Please check if the PR fulfills these requirements

- [ ] The commit message follows our [guidelines](https://github.com/oasis-engine/engine/blob/main/.github/COMMIT_MESSAGE_CONVENTION.md)
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)
#377 CharacterController from PhysX
### Other information:

### CODE REVIEW

File: packages/physics-physx/src/shape/index.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export * from "./PhysXBoxColliderShape";
+export * from "./PhysXSphereColliderShape";
+export * from "./PhysXCapsuleColliderShape";
+export * from "./PhysXColliderShape";
```
Comment (GuoLei1990) [inline packages/physics-physx/src/shape/index.ts:4]: Delete this file

File: packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+   */
+  resize(height: number) {
+    this._pxController.resize(height);
+  }
+
+  /**
+   * {@inheritDoc ICharacterController.updateShape }
+   */
+  updateShape(): void {
+    const controller = this._pxController;
```
Comment (GuoLei1990) [inline packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts:108]: Write this line into `if (shape._isDirty)`

File: packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    const shape = this._shape;
+    if (shape._isDirty) {
+      if (shape instanceof PhysXBoxColliderShape) {
```
Comment (GuoLei1990) [inline packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts:111]: cache a flag _isBoxShape, instanceof is very slow

File: packages/core/src/physics/CharacterController.ts [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+  /// Character is colliding to the sides.
+  COLLISION_SIDES = 1,
+  /// Character has collision above.
+  COLLISION_UP = 2,
+  /// Character has collision below.
+  COLLISION_DOWN = 4
+}
+
+/**
+ * Base class for character controllers.
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:24]: Opt comments, is not  `Base class`

File: packages/core/src/physics/CharacterController.ts [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+ */
+export class CharacterController extends Collider {
+  /** @internal */
+  _index: number = -1;
+  _nativeCharacterController: ICharacterController;
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:29]: Add /** @internal */

File: packages/core/src/physics/PhysicsManager.ts [hunk 6]
```diff
@@ -138,6 +143,13 @@ export class PhysicsManager {
     this._nativePhysicsManager.setGravity(gravity);
   }
 
+  /**
+   * The character controller manager.
+   */
+  get characterControllerManager(): ICharacterControllerManager {
```
Comment (GuoLei1990) [inline packages/core/src/physics/PhysicsManager.ts:149]: Delete this

File: packages/design/src/physics/characterkinematic/ICharacterControllerManager.ts [hunk 7]
```diff
@@ -0,0 +1,56 @@
+import { Vector3 } from "@oasis-engine/math";
+import { ICharacterController } from "./ICharacterController";
+import { IColliderShape } from "../shape";
+
+/**
+ * Manages an array of character controllers.
+ */
+export interface ICharacterControllerManager {
```
Comment (GuoLei1990) [inline packages/design/src/physics/characterkinematic/ICharacterControllerManager.ts:8]: Delete this and add createController to IPhysXManager;

File: packages/core/src/physics/CharacterController.ts (hunk 4 above)
Comment (yangfengzzz) [reply packages/core/src/physics/CharacterController.ts:24]: done

File: packages/core/src/physics/CharacterController.ts (hunk 5 above)
Comment (yangfengzzz) [reply packages/core/src/physics/CharacterController.ts:29]: done

File: packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts (hunk 3 above)
Comment (yangfengzzz) [reply packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts:111]: done

File: packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts (hunk 2 above)
Comment (yangfengzzz) [reply packages/physics-physx/src/characterkinematic/PhysXCharacterController.ts:108]: done

File: packages/physics-physx/src/shape/index.ts (hunk 1 above)
Comment (yangfengzzz) [reply packages/physics-physx/src/shape/index.ts:4]: done

File: packages/design/src/physics/characterkinematic/ICharacterControllerManager.ts (hunk 7 above)
Comment (yangfengzzz) [reply packages/design/src/physics/characterkinematic/ICharacterControllerManager.ts]: done

File: packages/core/src/physics/PhysicsManager.ts (hunk 6 above)
Comment (yangfengzzz) [reply packages/core/src/physics/PhysicsManager.ts]: done

File: packages/core/src/physics/CharacterController.ts [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+
+  set nonWalkableMode(newValue: ControllerNonWalkableMode) {
+    this._nonWalkableMode = newValue;
+    this._nativeCharacterController.setNonWalkableMode(newValue);
+  }
+
+  /**
+   * The contact offset for the controller.
+   */
+  get contactOffset(): number {
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:49]: Write this to ColliderShape

File: packages/core/src/physics/CharacterController.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    const oldCollider = shape._collider;
+    if (oldCollider !== this) {
+      if (oldCollider) {
+        oldCollider.removeShape(shape);
+      }
+      // create controller first which will examine shape is proper.
+      this._nativeCharacterController = this.engine.physicsManager._createController(shape);
+      this.engine.physicsManager._addCollider(this);
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:112]: Add valid condition

File: packages/core/src/physics/CharacterController.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+  /**
+   * Remove a collider shape.
+   * @param shape - The collider shape.
+   * @override
+   */
+  removeShape(shape: ColliderShape): void {
+    const index = this._shapes.indexOf(shape);
+    if (index !== -1) {
+      this._shapes.splice(index, 1);
+      this.engine.physicsManager._removeColliderShape(shape);
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:129]: Add valid condition

File: packages/core/src/physics/CharacterController.ts [hunk 11]
```diff
@@ -0,0 +1,163 @@
...
+   * @internal
+   * @override
+   */
+  _onLateUpdate() {
+    let position = this.entity.transform.worldPosition;
+    this._nativeCharacterController.getPosition(position);
+    this.entity.transform.worldPosition = position;
+    this._updateFlag.flag = false;
+  }
+}
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:163]: override `_onEnbale`/`_onDisable`

File: packages/core/src/physics/PhysicsManager.ts [hunk 12]
```diff
@@ -318,6 +338,37 @@ export class PhysicsManager {
...
+  /**
+   * Creates a new character controller.
+   * @param shape The controllers shape
+   * @internal
+   */
+  _createController(shape: ColliderShape): ICharacterController {
+    return this._nativePhysicsManager.createController(shape._nativeShape);
+  }
+
+  callColliderOnUpdate(): void {
```
Comment (GuoLei1990) [inline packages/core/src/physics/PhysicsManager.ts:361]: add `_` and write @internal

File: packages/core/src/physics/PhysicsManager.ts [hunk 13]
```diff
@@ -318,6 +338,37 @@ export class PhysicsManager {
...
+    const elements = this._colliders._elements;
+    for (let i = this._colliders.length - 1; i >= 0; --i) {
+      elements[i]._onUpdate();
+    }
+  }
+
+  callColliderOnLateUpdate(): void {
```
Comment (GuoLei1990) [inline packages/core/src/physics/PhysicsManager.ts:368]: add `_` and write @internal

File: packages/core/src/physics/enums/ControllerCollisionFlag.ts [hunk 14]
```diff
@@ -0,0 +1,11 @@
+/**
+ * The up axis of the collider shape.
+ */
+export enum ControllerCollisionFlag {
+  /// Character is colliding to the sides.
+  Collision_Sides = 1,
```
Comment (GuoLei1990) [inline packages/core/src/physics/enums/ControllerCollisionFlag.ts:6]: Write `Sides`、`Up`、`Down` is enough

File: packages/design/src/physics/IPhysicsManager.ts [hunk 15]
```diff
@@ -36,6 +37,24 @@ export interface IPhysicsManager {
...
+   * Remove ICharacterController.
+   * @param characterController The Character Controller.
+   */
+  removeCharacterController(characterController: ICharacterController): void;
+
+  /**
+   * Creates a new character controller.
+   * @param shape The controllers shape
+   */
+  createController(shape: IColliderShape): ICharacterController;
```
Comment (GuoLei1990) [inline packages/design/src/physics/IPhysicsManager.ts:56]: createCharacterController

File: packages/core/src/physics/CharacterController.ts (hunk 8 above)
Comment (yangfengzzz) [reply packages/core/src/physics/CharacterController.ts]: done

File: packages/core/src/physics/CharacterController.ts (hunk 10 above)
Comment (yangfengzzz) [reply packages/core/src/physics/CharacterController.ts:118]: done

File: packages/core/src/physics/CharacterController.ts (hunk 9 above)
Comment (yangfengzzz) [reply packages/core/src/physics/CharacterController.ts]: done

File: packages/core/src/physics/PhysicsManager.ts (hunk 13 above)
Comment (yangfengzzz) [reply packages/core/src/physics/PhysicsManager.ts:380]: done

File: packages/core/src/physics/enums/ControllerCollisionFlag.ts (hunk 14 above)
Comment (yangfengzzz) [reply packages/core/src/physics/enums/ControllerCollisionFlag.ts:6]: done

File: packages/core/src/physics/enums/ControllerNonWalkableMode.ts [hunk 16]
```diff
@@ -0,0 +1,9 @@
+/**
+ * The up axis of the collider shape.
+ */
+export enum ControllerNonWalkableMode {
+  /// Stops character from climbing up non-walkable slopes, but doesn't move it otherwise
+  Prevent_Climbing,
```
[the remark this answers is not in the document]
Comment (yangfengzzz) [reply packages/core/src/physics/enums/ControllerNonWalkableMode.ts:6]: done

File: packages/core/src/physics/CharacterController.ts [hunk 17]
```diff
@@ -0,0 +1,179 @@
...
+ */
+export class CharacterController extends Collider {
+  /** @internal */
+  _index: number = -1;
+  /** @internal */
+  _nativeCharacterController: ICharacterController;
+
+  private _stepOffset: number = 0;
+  private _nonWalkableMode: ControllerNonWalkableMode = ControllerNonWalkableMode.Prevent_Climbing;
+  private _contactOffset: number = 0;
```
Comment (GuoLei1990) [inline packages/core/src/physics/CharacterController.ts:18]: Delete this

File: packages/physics-physx/src/shape/PhysXColliderShape.ts [hunk 18]
```diff
@@ -108,13 +113,16 @@ export abstract class PhysXColliderShape implements IColliderShape {
     this._pxShape.setLocalPose(transform);
   }
 
-  protected _allocShape(material: PhysXPhysicsMaterial): void {
+  protected _initialize(material: PhysXPhysicsMaterial, index: number): void {
```
Comment (GuoLei1990) [inline packages/physics-physx/src/shape/PhysXColliderShape.ts:116]: `id` is better than `index`

File: packages/design/src/physics/IPhysicsManager.ts (hunk 15 above)
Comment (yangfengzzz) [reply packages/design/src/physics/IPhysicsManager.ts:56]: done

File: packages/core/src/physics/CharacterController.ts (hunk 17 above)
Comment (yangfengzzz) [reply packages/core/src/physics/CharacterController.ts:18]: done

**Review by GuoLei1990 — APPROVED** [review]

**Pull request merged**